### PR TITLE
Added example for CPP Dll useage

### PR DESCRIPTION
**Pull request by Hecatron**
### What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Docs
- [ ] New Binding issue #___
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [X] Other, please describe:

This is basically an example of using tauri within a dll library (windows).
So that the webview is generated / run from within the dll.

The idea being you can run / call the dll from an existing C / C++ project for interop with the project.

### Does this PR introduce a breaking change?
Nope

### CODE REVIEW

**Comment by nothingismagick** [comment]
Really, really interesting!

**Comment by Hecatron** [comment]
I was hoping to use this method to write a freecad extension which normally uses python / dll's normally written in c++
the idea being to write a rust / tarui dll library that could be loaded instead.

**Comment by nothingismagick** [comment]
Amazing

File: examples/cpp-dll/src-app1/src/main.rs
Comment (amrbashir) [inline examples/cpp-dll/src-app1/src/main.rs]: use `library_filname()` function for cross-platform support
```suggestion
use libloading::{Library, library_filename, Symbol};
type LibFunctionType1 = fn();

fn main() {
  let library_path = library_filename("../src-dll1/target/debug/dll1");
  println!("Loading lib_test1() from {}", library_path);

  unsafe {
    let lib = Library::new(library_path).unwrap();
```

File: examples/cpp-dll/src-app1/src/main.rs
Comment (amrbashir) [inline examples/cpp-dll/src-app1/src/main.rs]: Could use a better name
```suggestion
    let run_tauri: Symbol<LibFunctionType1> = lib.get(b"run_tauri").unwrap();
```

File: examples/cpp-dll/src-dll1/Cargo.toml
Comment (amrbashir) [inline examples/cpp-dll/src-dll1/Cargo.toml]: Also could use a better name, and the directory too should be called `src-tauri`
```suggestion
name = "tauri-app"
```

File: examples/cpp-dll/src-dll1/src/lib.rs
Comment (amrbashir) [inline examples/cpp-dll/src-dll1/src/lib.rs]:
```suggestion
pub extern fn run_tauri() {
```

**Review by amrbashir — CHANGES_REQUESTED** [review]
Thank you for this cool example

File: examples/cpp-dll/src-dll1/Cargo.toml
Comment (Hecatron) [reply examples/cpp-dll/src-dll1/Cargo.toml]: I've used tauri_app just to follow the snake_case as that's what the build outputs

File: examples/cpp-dll/src-dll1/src/lib.rs
Comment (Hecatron) [reply examples/cpp-dll/src-dll1/src/lib.rs]: Now renamed

File: examples/cpp-dll/src-app1/src/main.rs
Comment (Hecatron) [reply examples/cpp-dll/src-app1/src/main.rs]: Now renamed

File: examples/cpp-dll/src-app1/src/main.rs
Comment (Hecatron) [reply examples/cpp-dll/src-app1/src/main.rs]: Thanks for this one, I wanted to do it this way but didn't know about library_filename

**Comment by Hecatron** [comment]
Let me know if anything else needs changing

**Review by amrbashir — CHANGES_REQUESTED** [review]
A couple more things to change: 
- the example directory name from `cpp-dll` to `tauri-dynamic-lib`
- the read me in this example should have the name changes you just made, and it should not be focused on Windows dll or C/C++. The wording should be more general, it doesn't matter which language consumes this dynamic lib nor what OS it is on.

**Comment by Hecatron** [comment]
Ok made some more changes, let me know if you spot anything else

File: examples/tauri-dynamic-lib/Readme.md
Comment (amrbashir) [inline examples/tauri-dynamic-lib/Readme.md:3]:
```suggestion
This is an example of compiling tauri as a dynamic shared library and running it from another app.
```

Comment (amrbashir) [inline examples/tauri-dynamic-lib/Readme.md]:
```suggestion
This is an example of compiling tauri as a dynamic shared library and running it from another app.
```

File: examples/tauri-dynamic-lib/Readme.md
Comment (amrbashir) [inline examples/tauri-dynamic-lib/Readme.md]:
```suggestion
  * src-app1 is a small example of calling tauri from a dynamic shared library through FFI.
```

**Review by amrbashir — CHANGES_REQUESTED** [review]

File: examples/tauri-dynamic-lib/Readme.md
Comment (Hecatron) [reply examples/tauri-dynamic-lib/Readme.md]: now updated

File: examples/tauri-dynamic-lib/Readme.md
Comment (Hecatron) [reply examples/tauri-dynamic-lib/Readme.md]: now updated

**Comment by lucasfernog** [comment]
Maybe the example could also show how to interop with C code?

**Comment by Hecatron** [comment]
> Maybe the example could also show how to interop with C code?

It may be different based on which platform you're using which might complicate things.
(which is why I did it in rust as that has a lot of polish / abstracts away the implementation)
Typically in real life you'd probably use something like https://github.com/eqrion/cbindgen
or https://github.com/rust-lang/rust-bindgen or https://cxx.rs/ for the interop to generate the needed headers

But basically all you're doing from the C side is calling a function in a dll or shared library.
But with C there's lots of different ways to do the same thing based on the platform and libs in use.

What I was more interested in to be honest was if this sort of implementation would mess up the bundling or the rust build side of things since tauri uses custom build scripts, fortunately it seems to all just work which is nice.

**Review by amrbashir — APPROVED** [review]
LGTM! but I will wait for Lucas to approve before merge.

**Pull request merged**
